[#182] Add support for GenQuery2 to genquery.py module (4-3-stable) (#206)
irods/irods_rule_engine_plugin_python: README.md

  * *options* is a bitmask of extra options, and defaults to a value of 0. `genquery.Option.NO_DISTINCT` is one such extra option, as is RETURN_TOTAL_ROW_COUNT (although in the latter case, using the `Query` object's `row_count` method should be preferred.)
  * *parser* defines which GenQuery parser to use for querying the catalog. The following values are supported:
    - `Parser.GENQUERY1` configures the `Query` object to use GenQuery1 (i.e. the traditional parser). This is the default.
    - `Parser.GENQUERY2` configures the `Query` object to use GenQuery2. GenQuery2 is an experimental parser with several enhancements over GenQuery1.
      - When using the GenQuery2 parser, the following applies:
        - The `output` constructor parameter is ignored.
        - The `case_sensitive` constructor parameter is ignored.
        - The `options` constructor parameter is ignored.
        - The `total_rows` member function of the `Query` class always returns `None`.

>   * *order_by* is a string holding the sorting instructions for a GenQuery2 query. Defaults to an empty string. Only recognized by the GenQuery2 parser.

Author: Which options are you referring to?

`order_by` accepts any string as long as it follows the SQL rules for the ORDER-BY clause.

>   * *order_by* is a string holding the sorting instructions for a GenQuery2 query. Defaults to an empty string. Only recognized by the GenQuery2 parser.

Reviewer: if that is 'known' then it doesn't need to go here...   but... maybe a mention of 'standard SQL' or something.

>   * *order_by* is a string holding the sorting instructions for a GenQuery2 query. Defaults to an empty string. Only recognized by the GenQuery2 parser.

Author: Sure. Will adjust.

>   * *order_by* is a string holding the sorting instructions for a GenQuery2 query. Defaults to an empty string. Only recognized by the GenQuery2 parser.

Author: Done.

> When a GenQuery2 resultset is done being processed, it is best practice to call the `close()` member function. Doing this will instruct the server to immediately free any resources allocated to the `Query` object. This is extremely important when multiple `Query` objects are executed within a single rule.

Author: The use of word, `processing`, is little weird to me.

What are your thoughts on the following?
```
When a GenQuery2 resultset is no longer needed, ...
```
Or
```
When processing of a GenQuery2 resultset is complete, ...
```

> When a GenQuery2 resultset is done being processed, it is best practice to call the `close()` member function. Doing this will instruct the server to immediately free any resources allocated to the `Query` object. This is extremely important when multiple `Query` objects are executed within a single rule.

Reviewer: ah,, yes....    

`When the processing of a GenQuery2 resultset is complete ...`

> When a GenQuery2 resultset is done being processed, it is best practice to call the `close()` member function. Doing this will instruct the server to immediately free any resources allocated to the `Query` object. This is extremely important when multiple `Query` objects are executed within a single rule.

Author: Done.

  * *case-sensitive* is normally `True`. If it is set to `False`, the condition string will be uppercased, and the query will be executed without regard to case.  This allows for more permissive matching on the names of resources, collections, data objects, etc.
  * *options* is a bitmask of extra options, and defaults to a value of 0. `genquery.Option.NO_DISTINCT` is one such extra option, as is RETURN_TOTAL_ROW_COUNT (although in the latter case, using the `Query` object's `row_count` method should be preferred.)

> As of iRODS 4.3.3, the following parameters are optional:

Reviewer: > Or, just get rid of this mention of the version entirely and include this parameter in the above list of optional parameters.

just including in the list as optional above seems sufficient... (perhaps with a 'since 4.3.3.0' note?)

> As of iRODS 4.3.3, the following parameters are optional:

Author: Will update

> As of iRODS 4.3.3, the following parameters are optional:

Author: Done

        - The `case_sensitive` constructor parameter is ignored.
        - The `options` constructor parameter is ignored.
        - The `total_rows` member function of the `Query` class always returns `None`.
  * *order_by* is a string holding the sorting instructions for a GenQuery2 query. Defaults to an empty string. Only used by the GenQuery2 parser.

> As a best practice, when using GenQuery2, prefer calling the `close()` member function when done processing the resultset. Doing this will instruct the server to immediately free any resources allocated to the `Query` object. This is extremely important when multiple `Query` objects are executed within a single rule.

Reviewer 2: I find the "prefer" and "extremely important" juxtaposition a little awkward here. How does this wording look?
```suggestion
When a GenQuery2 resultset is done being processed, it is best practice to call the `close()` member function. Doing this will instruct the server to immediately free any resources allocated to the `Query` object. This is extremely important when multiple `Query` objects are executed within a single rule.
```

Author: Done

  "server_zone": "tempZone"
}
```

## How do I escape embedded single quotes using the Query class and GenQuery2?

GenQuery2 provides two ways to escape embedded single quotes. They are as follows:

> - Use a single quote to escape a single quote as defined by the SQL standard

Reviewer: can we see an example of this first format as well?

Author: Done.

- Or, use the hexadecimal encoding

Below is an example showing how hexadecimal can be used to escape embedded single quotes.

> ```python

Reviewer: show the string that is being embedded... I think it is `'_quotes_around_me_'`

Author: Will do.

Author: Done.

from genquery import Query, Parser
row = Query(callback, 'DATA_NAME', "DATA_NAME = '\\x27_quotes_around_me_\\x27'", parser=Parser.GENQUERY2).first()
callback.writeLine('serverLog', f'row = {row}')
```

Notice the use of `\\x` rather than `\x`. This is required to keep Python from interpreting the hexadecimal value.

## How do I sort data using the Query class and GenQuery2?

Use the `order_by` parameter. Two examples are provided below.

> Assume we have five collections and we want to sort them by the length of their logical path, but in descending order.

Reviewer: perhaps a third example with data object size?  Does it know it's a number?  Does it sort 12, 5, and 321 correctly?

Author: Sorting is handled by the database. My guess is the results can vary depending on the configuration of the database _(i.e. organizations can use different collations)_.

Here's what I get with `iquery`.
```bash
$ ils -l
/tempZone/home/kory:
  kory              0 demoResc            0 2024-07-22.14:01 & 0.txt
  kory              0 demoResc          321 2024-07-23.20:35 & 321.bin
  kory              0 demoResc            6 2024-07-23.20:33 & bar
  kory              0 demoResc           29 2024-07-23.20:34 & cats
  kory              0 demoResc            0 2024-07-17.16:53 & foo
  kory              0 demoResc            0 2024-07-19.21:11 & thinger
$ iquery "select DATA_SIZE order by DATA_SIZE" | jq
[
  [
    "0"
  ],
  [
    "6"
  ],
  [
    "29"
  ],
  [
    "321"
  ]
]
$ iquery "select DATA_SIZE order by DATA_SIZE desc" | jq
[
  [
    "321"
  ],
  [
    "29"
  ],
  [
    "6"
  ],
  [
    "0"
  ]
]
```
`iquery` appears to handle `DATA_SIZE` just fine.

Curious ... what does an example using `DATA_SIZE` offer over the current examples?

Reviewer: oh good - it knows they are ... integers/numbers somehow.

some people don't realize they can get more than just COLL_NAME, DATA_NAME - no real requirement to use DATA_SIZE.

Author: GenQuery2 also supports casting.

We have an open issue for documenting all of this in irods/irods_docs. I just haven't started writing it yet.

Reviewer: yep, no worries - resolving.
